### PR TITLE
Nolock Option for TSDB CLI Tool

### PR DESCRIPTION
**Pull request by glightfoot**
Adds an option to not lock the TSDB when opening. Might need some additional warnings, or maybe just made default, but this seems to work fine on my internal testing against a heavily used 2.7.1 prometheus server.

fixes https://github.com/prometheus/tsdb/pull/517

### CODE REVIEW

File: cmd/tsdb/main.go [hunk 1]
```diff
@@ -55,6 +57,13 @@ func main() {
 		analyzeLimit         = analyzeCmd.Flag("limit", "how many items to show in each list").Default("20").Int()
 	)
 
+	options := &tsdb.Options{
+		WALSegmentSize:    wal.DefaultSegmentSize,
+		RetentionDuration: 15 * 24 * 60 * 60 * 1000, // 15 days in milliseconds
+		BlockRanges:       tsdb.ExponentialBlockRanges(int64(2*time.Hour)/1e6, 3, 5),
+		NoLockfile:        false,
```
Comment (krasi-georgiev) [inline cmd/tsdb/main.go]: since this is set as a global option I expect this to be set here
```suggestion
		NoLockfile:        *noLock,
```

**Comment by brian-brazil** [comment]
Rather than the user having to know of this magic flag, could we verify where it's safe and do it automatically?

**Comment by glightfoot** [comment]
@brian-brazil I like that more, but don't know enough about the workings to know if it's safe 100% of the time or not. From my browsing, it looks like it's only needed for `ls` and `analyze` - if those are guaranteed to be safe running against a running prometheus, I can just hard code it and remove the flag

**Comment by brian-brazil** [comment]
The question is if there's any potentially mutating actions, such as migrating the WAL or compactions.

**Comment by krasi-georgiev** [comment]
The only problem I see when running on a live DB could be incorrect results if some compaction happens during the list/analyse command or under windows the compaction might fail to delete the block when it is open for reading.

Either way I think the suggestion to make it a default and non configurable is fine and can a short warning like:
`listCmd              = cli.Command("ls", "list db blocks (running on a live DB can have unpredicted results).`
analyzeCmd           = cli.Command("analyze", "analyze churn, label pair cardinality (running on a live DB can have unpredicted results).")

**Comment by brian-brazil** [comment]
That seems dangerous, a read only Open method that didn't load the WAL would be safer.

**Comment by krasi-georgiev** [comment]
Aaaah , now I understand what you mean I completely forgot that opening the DB also loads the WAL file.

Once we have the [noop wal PR](https://github.com/prometheus/tsdb/pull/519) merged this would be possible.

**Comment by glightfoot** [comment]
Ok, I'll wait until #519 is merged and then modify to use that instead

**Comment by brian-brazil** [comment]
There's a bit more to it than that, compaction and repair also needs disabling.

**Comment by krasi-georgiev** [comment]
compaction can be disabled with `db.DisableCompactions()`.

@brian-brazil are you referring to WAL repair? This is also disabled when you pass a nil WAL.

**Comment by brian-brazil** [comment]
No, the index repair. DisableCompactions would also be racy.

**Comment by glightfoot** [comment]
What about adding a ReadOnly field to the db.Options struct? Or would you rather have individual options for disabling specific features (NopWal, NoCompact, NoRepair, etc) and add an OpenReadOnly method to the db? I'm happy to do the work if we can identify all the areas that need changing

**Comment by brian-brazil** [comment]
A ReadOnly option sounds good to me.

**Comment by krasi-georgiev** [comment]
ok lets try with the ReadOnly option when opening a DB to see how it would look like in code.

**Comment by glightfoot** [comment]
@krasi-georgiev - Roughly like that. Still needs #519 merged before testing and finishing up the work on it

**Comment by krasi-georgiev** [comment]
yep it doesn't look too bad. Would also need a test to ensure that opening a db with ReadOnly does indeed disable all mutating actions.

For example 
- open the db with a corrupt index and ensure it is not repaired (testdata/repair_index_version)
- Ensure that the compaction is disabled - 
```
testutil.Equals(t, 0.0, prom_testutil.ToFloat64(db.metrics.compactionsTriggered,"...")
testutil.Equals(t, 0.0, prom_testutil.ToFloat64(db.metrics.compactionsSkipped,"...")

db.compactc <- struct{}{}
testutil.Equals(t, 1.0, prom_testutil.ToFloat64(db.metrics.compactionsTriggered,"...")
testutil.Equals(t, 1.0, prom_testutil.ToFloat64(db.metrics.compactionsSkipped, "...")
```

- Ensure that no wal is created.

**Comment by krasi-georgiev** [comment]
this other [PR](https://github.com/prometheus/tsdb/pull/532) reminded me that the size/time retention is another thing that need to be disabled

**Comment by krasi-georgiev** [comment]
ping @glightfoot

**Comment by krasi-georgiev** [comment]
@glightfoot 	I hope you don't mind, but since I don't see any activity here will continue the work in https://github.com/prometheus/tsdb/pull/541

**Pull request closed without merging**